source/grid/grid_generator.cc

```diff
@@ -7090,6 +7090,273 @@ namespace GridGenerator
 
 
 
+  template <int dim, int spacedim>
+  void
+  convert_hypercube_to_simplex_mesh(const Triangulation<dim, spacedim> &in_tria,
+                                    Triangulation<dim, spacedim> &out_tria)
+  {
+    AssertDimension(in_tria.n_global_levels(), 1);
+
+    Assert(dim > 1, ExcNotImplemented());
+
+    std::vector<Point<spacedim>> vertices;
+    std::vector<CellData<dim>>   cells;
+    SubCellData                  subcell_data;
+
+    // store for each vertex and face the assigned index so that we only
+    // assign them a value only once
+    std::vector<unsigned int> indices_vertics(in_tria.n_vertices(),
+                                              numbers::invalid_unsigned_int);
+    std::vector<unsigned int> indices_faces(in_tria.n_faces(),
+                                            numbers::invalid_unsigned_int);
+
+    // temporary array storing the global indices of each cell entity in the
+    // sequence: vertices, edges/faces, cell
+    std::array<unsigned int, dim == 2 ? 9 : 14> indices_local;
```

Comment:
You don't need to know this: `std::array` objects are allocated on the stack, not the heap. So there is no cost to moving the declaration of this variable into the loop over all cells. Doing so makes it clear that (i) this variable does not carry state from one loop iteration to the next, (ii) it's clear that the variable should not be used after the end of the loop body. Both of these things are quite useful when reading through code at a later point.

Separately, the variable stores *vertex* indices. Maybe name it `local_vertex_indices`?